src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -305,6 +305,24 @@ private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WI
             
             Debug.Assert(state != null, "OnRequestError: state is null");
 
+            Debug.Assert(!(unchecked((int)asyncResult.dwError) == Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER ||
+                unchecked((int)asyncResult.dwError) == unchecked((int)0x80090321)), // SEC_E_BUFFER_TOO_SMALL
+                $"Unexpected async error in WinHttpRequestCallback: {unchecked((int)asyncResult.dwError)}, WinHttp API: {unchecked((uint)asyncResult.dwResult.ToInt32())}");
+
+#if DEBUG
+            // Add instrumentation to catch unexpected errors and fail fast to produce a crash dump. Issue #7812.
+            // If the failure is not caused by the above error codes, need to get more information from asyncResult.dwResult.
+            // There are only two cases in the below switch statement which explicitly set WinHttpException.
+            if (unchecked((uint)asyncResult.dwResult.ToInt32()) == Interop.WinHttp.API_SEND_REQUEST ||
+                ((unchecked((uint)asyncResult.dwResult.ToInt32()) == Interop.WinHttp.API_RECEIVE_RESPONSE) &&
+                (asyncResult.dwError != Interop.WinHttp.ERROR_WINHTTP_RESEND_REQUEST) &&
+                (asyncResult.dwError != Interop.WinHttp.ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED) &&
+                (asyncResult.dwError != Interop.WinHttp.ERROR_WINHTTP_OPERATION_CANCELLED)))
+            {
+                Debug.Fail($"Unexpected error: {unchecked((int)asyncResult.dwError)}, WinHttp API: {unchecked((uint)asyncResult.dwResult.ToInt32())}");
```

Comment:
Why does this string just say this: "Unexpected error:" when the string above it says "Unexpected async error in WinHttpRequestCallback:"?

Same comment for above for the new code you added in this latest commit.